x/lightclient/ante/ibc_msg_update_client.go

```diff
@@ -1,82 +1,153 @@
+	}
+	if err != nil {
+		return err
+	}
+
+	// ~~~~~
+	// now we know that the msg is a header, and it was produced by a sequencer
+	// ~~~~~
+
+	if !seq.Bonded() {
+		// we assume here that sequencers will not propose blocks on other chains connected to the hub except for their rollapp
+		return gerrc.ErrInvalidArgument.Wrap("header is from unbonded sequencer")
+	}
+
+	rollapp, ok := i.raK.GetRollapp(ctx, seq.RollappId)
+	if !ok {
+		return gerrc.ErrInternal.Wrapf("get rollapp from sequencer: rollapp: %s", seq.RollappId)
+	}
+
+	// TODO: in hard fork will need to also use revision to make sure not from old revision
+	h := header.GetHeight().GetRevisionHeight()
+	stateInfos, err := i.getStateInfos(ctx, rollapp.RollappId, h)
+	if err != nil {
+		return errorsmod.Wrap(err, "get state infos")
+	}
+
+	if stateInfos.containingHPlus1 != nil {
+		// the header is pessimistic: the state update has already been received, so we check the header doesn't mismatch
+		return errorsmod.Wrap(i.validateUpdatePessimistically(ctx, stateInfos, header.ConsensusState(), h), "validate pessimistic")
+	}
+
+	// the header is optimistic: the state update has not yet been received, so we save optimistically
+	return errorsmod.Wrap(i.k.SaveSigner(ctx, seq.Address, msg.ClientId, h), "save updater")
+}
+
+var (
+	errIsMisbehaviour   = errorsmod.Wrap(gerrc.ErrFailedPrecondition, "misbehavior evidence is disabled for canonical clients")
+	errNoHeader         = errors.New("message does not contain header")
+	errProposerMismatch = errorsmod.Wrap(gerrc.ErrInvalidArgument, "validator set proposer not equal header proposer field")
+)
+
+func (i IBCMessagesDecorator) getSequencer(ctx sdk.Context, header *ibctm.Header) (sequencertypes.Sequencer, error) {
+	proposerBySignature := header.ValidatorSet.Proposer.GetAddress()
+	proposerByData := header.Header.ProposerAddress
+	// Does ibc already guarantee this equal to header.ProposerAddr? I don't think so
+	if !bytes.Equal(proposerBySignature, proposerByData) {
+		return sequencertypes.Sequencer{}, errProposerMismatch
 	}
+	return i.k.SeqK.SequencerByDymintAddr(ctx, proposerByData)
+}
 
+func getHeader(msg *ibcclienttypes.MsgUpdateClient) (*ibctm.Header, error) {
 	clientMessage, err := ibcclienttypes.UnpackClientMessage(msg.ClientMessage)
 	if err != nil {
-		return nil
+		return nil, errorsmod.Wrap(err, "unpack client message")
 	}
-	_, ok = clientMessage.(*ibctm.Misbehaviour)
+	_, ok := clientMessage.(*ibctm.Misbehaviour)
 	if ok {
-		return errorsmod.Wrap(gerrc.ErrFailedPrecondition, "misbehavior evidence is disabled for canonical clients")
+		return nil, errIsMisbehaviour
 	}
 	header, ok := clientMessage.(*ibctm.Header)
 	if !ok {
-		return nil
+		return nil, errNoHeader
 	}
+	return header, nil
+}
+
+// if containingHPlus1 is not nil then containingH also guaranteed to not be nil
+type stateInfos struct {
+	containingH      *rollapptypes.StateInfo
+	containingHPlus1 *rollapptypes.StateInfo
```

Comment:
`containingHPlus1` used to get next sequencer
can we take the `NextSequencer` field from the `containingH` state info instead? (after the HF PR)

Reply:
sure